src/lib/statistics/chunk_statistics/histograms/abstract_histogram.cpp

```diff
@@ -0,0 +1,620 @@
+  return next_value(value, _supported_characters);
+}
+
+template <typename T>
+float AbstractHistogram<T>::_bin_share(const BinID bin_id, const T value) const {
+  return static_cast<float>(value - _bin_min(bin_id)) / _bin_width(bin_id);
+}
+
+template <>
+float AbstractHistogram<std::string>::_bin_share(const BinID bin_id, const std::string value) const {
+  /**
+   * Returns the share of values smaller than `value` in the given bin.
+   *
+   * We need to convert strings to their numerical representation to calculate a share.
+   * This conversion is done based on prefixes because strings of arbitrary length cannot be converted to a numerical
+   * representation that satisfies the following requirements:
+   *  1. For two strings s1 and s2: s1 < s2 -> repr(s1) < repr(s2)
+   *  2. For two strings s1 and s2: dist(s1, s2) == repr(s2) - repr(s1)
+   *  repr(s) is the numerical representation for a string s, and dist(s1, s2) returns the number of strings between
+   *  s1 and s2 in the domain of strings with at most length `string_prefix_length`
+   *  and the set of supported characters `supported_characters`.
+   *
+   * Thus, we calculate the range based only on a domain of strings with a maximum length of `string_prefix_length`
+   * characters.
+   * However, we make use of a trick: if the bin edges share a common prefix, we strip that common prefix and
+   * take the substring starting after that prefix.
+   *
+   * Example:
+   *  - bin: ["intelligence", "intellij"]
+   *  - supported_characters: [a-z]
+   *  - string_prefix_length: 4
+   *  - value: intelligent
+   *
+   *  Traditionally, if we did not strip the common prefix, we would calculate the range based on the
+   *  substring of length `string_prefix_length`, which is "inte" for both lower and upper edge of the bin.
+   *  We could not make a reasonable assumption how large the share is.
+   *  Instead, we strip the common prefix ("intelli") and calculate the share based on the numerical representation
+   *  of the substring after the common prefix.
+   *  That is, what is the share of values smaller than "gent" in the range ["gence", "j"]?
+   */
+  const auto bin_min = _bin_min(bin_id);
+  const auto bin_max = _bin_max(bin_id);
+  const auto common_prefix_len = common_prefix_length(bin_min, bin_max);
+
+  DebugAssert(value.substr(0, common_prefix_len) == bin_min.substr(0, common_prefix_len),
+              "Value does not belong to bin");
+
+  const auto value_repr = _convert_string_to_number_representation(value.substr(common_prefix_len));
+  const auto min_repr = _convert_string_to_number_representation(bin_min.substr(common_prefix_len));
+  const auto max_repr = _convert_string_to_number_representation(bin_max.substr(common_prefix_len));
+  return static_cast<float>(value_repr - min_repr) / (max_repr - min_repr + 1);
+}
+
+template <typename T>
+bool AbstractHistogram<T>::_can_prune(const PredicateCondition predicate_type, const AllTypeVariant& variant_value,
```

Comment:
Should a histogram know what "pruning" means or is it just a storage entity? I would expect some entity on a higher abstraction layer to translate from "the histogram has these values" to "this predicate can be used for pruning"

Reply:
I guess solved below?